tests/system/apmserver.py

```diff
@@ -27,43 +30,47 @@ class BaseTest(TestCase):
 
     def setUp(self):
         super(BaseTest, self).setUp()
-        if diagnostic_interval > 0:
-            self.diagnostics_path = os.path.join(self.working_dir, "diagnostics")
-            os.makedirs(self.diagnostics_path)
-            self.running = True
-            self.diagnostic_thread = threading.Thread(
-                target=self.dump_diagnotics, kwargs=dict(interval=diagnostic_interval))
-            self.diagnostic_thread.daemon = True
-            self.diagnostic_thread.start()
-
-    def tearDown(self):
-        if diagnostic_interval > 0:
-            self.running = False
-            self.diagnostic_thread.join(timeout=30)
-        super(BaseTest, self).tearDown()
+        self.setup_diagnostics()
+
+    def setup_diagnostics(self):
+        if diagnostic_interval <= 0:
+            return
+        self.addCleanup(self.cleanup_diagnostics)
+        self.diagnostics_path = os.path.join(self.working_dir, "diagnostics")
+        os.makedirs(self.diagnostics_path)
+        self.running = True
+        self.diagnostic_thread = threading.Thread(
+            target=self.dump_diagnotics, kwargs=dict(interval=diagnostic_interval))
+        self.diagnostic_thread.daemon = True
+        self.diagnostic_thread.start()
+
+    def cleanup_diagnostics(self):
+        self.running = False
+        self.diagnostic_thread.join(timeout=30)
+
+    def dump_diagnotics(self, interval=2):
+        while self.running:
+            time.sleep(interval)
```

Comment:
Can you please either replace this `time.sleep` with a [threading.Timer](https://docs.python.org/3/library/threading.html#timer-objects), or add a TODO? This existing code will slow tearDown, as the sleep won't be interrupted.

Reply:
Thanks for the suggestion. This code hasn't changed here, so I only added a TODO for now. Using `threading.Timer` would be a bit more complicated from what I saw and I'd like to move on with these changes.